packages/devtools_shared/lib/src/extensions/extension_manager.dart

```diff
@@ -0,0 +1,168 @@
+/// `<parent_package_root>/extension/devtools/`.
+const extensionBuildDefault = 'build';
+
+class ExtensionsManager {
+  ExtensionsManager({required this.buildDir});
+
+  final String buildDir;
+
+  String get servedExtensionsPath => path.join(buildDir, extensionRequestPath);
+
+  final devtoolsExtensions = <DevToolsExtensionConfig>[];
+
+  void serveAvailableExtensions(String? rootPath) {
+    devtoolsExtensions.clear();
+
+    if (rootPath != null) {
+      // TODO(kenz): use 'findExtensions' from package:extension_discovery once it
+      // is published.
+      // final extensions = findExtensions(
+      //   'devtools',
+      //   packageConfig: '$rootPath/.dart_tool/package_config.json',
+      // );
+      final extensions = <Extension>[];
+      for (final extension in extensions) {
+        final config = extension.config;
+        if (config is! Map) {
+          // Fail gracefully. Invalid content in the extension's config.json.
+          continue;
+        }
+        final configAsMap = config as Map<String, Object?>;
+
+        // This should be relative to the 'extension/devtools/' directory and
+        // defaults to 'build';
+        final relativeExtensionLocation =
+            configAsMap['buildLocation'] as String? ?? 'build';
+
+        // TODO(kenz): this is hacky. Unclear if this will work for windows.
+        var rootWithoutFile = extension.rootUri.toString();
+        if (rootWithoutFile.startsWith('file:///Users')) {
+          rootWithoutFile = rootWithoutFile.replaceFirst('file:///', '/');
+        }
+
+        final location = path.join(
+          rootWithoutFile,
+          extensionLocation,
+          relativeExtensionLocation,
+        );
+
+        try {
+          final pluginConfig = DevToolsExtensionConfig.parse({
+            ...configAsMap,
+            DevToolsExtensionConfig.pathKey: location,
+          });
+          devtoolsExtensions.add(pluginConfig);
+        } on StateError catch (e) {
+          print(e.message);
+          continue;
+        }
+      }
+    }
+
+    _resetServedPluginsDir();
+    for (final extension in devtoolsExtensions) {
+      _moveToServedExtensionsDir(extension.name, extension.path);
+    }
+  }
+
+  void _resetServedPluginsDir() {
+    final buildDirectory = Directory(buildDir);
+    if (!buildDirectory.existsSync()) {
+      throw Exception('The build directory does not exist.');
+    }
+
+    // Destroy and recreate the 'devtools_extensions' directory where extension
+    // assets are served.
+    final servedExtensionsDir = Directory(servedExtensionsPath);
+    if (servedExtensionsDir.existsSync()) {
+      servedExtensionsDir.deleteSync(recursive: true);
+    }
+    servedExtensionsDir.createSync();
+  }
+
+  void _moveToServedExtensionsDir(
+    String extensionPackageName,
+    String extensionPath,
+  ) {
+    final newExtensionPath =
+        path.join(servedExtensionsPath, extensionPackageName);
+    copyPathSync(extensionPath, newExtensionPath);
```

Comment:
In general, we should avoid using `fooSync()` APIs in production as they block the event loop for the duration of the call.